### DISCUSSION TITLE
Unable to read file from folder

### DISCUSSION THREAD

**Discussion by shabirjan** [pnp/pnpcore#1322] [Q&A] [answered]
Hi Everyone,

I am facing a weird issue while reading files from Sub folder in Documents folder using Pnp.Core. Below is the Code that I have tried till now.

               using(var pnpContext = await pnpContextFactory.CreateAsync("Default")) {
                var documentFolder = (await pnpContext.Web.Lists.GetByTitleAsync("Dokumenter", p => p.RootFolder));
                var allStoreFolders = (await  documentFolder.RootFolder.GetAsync(f => f.Folders)).Folders;
            
                foreach(var folder in allStoreFolders.AsRequested()) {

                    //  Console.WriteLine(folder.Name);
                    if (folder.Name == "BC10014")
                    {
                        Console.WriteLine("Folder Found");
                        await folder.LoadAsync(f => f.Files);

                        foreach (var file in folder.Files.AsRequested())
                        {
                            var extension = Path.GetExtension(file.Name);
                            if (extension == ".txt")
                            {
                                Console.WriteLine($"Filename: {file.ServerRelativeUrl}");

                            }
                        }
                    }
                } 

But it never returns the file from the sub folder. My site structure and folders are attached in the screenshot.  The only issue i think is the link for **Dokumenter** has empty space between, i.e  Delte%20dokumenter, but I don't know whether this can be the reason.

![image](https://github.com/pnp/pnpcore/assets/3787052/146c1447-29b1-4371-8794-49cdadbbb119)

<img width="713" alt="image" src="https://github.com/pnp/pnpcore/assets/3787052/dc2df7b5-e3d8-49d0-b89a-5bc61893cf2b">

**Comment by jansenbe** [comment]
@shabirjan: What about below approach:

```csharp
string sharedDocumentsFolderUrl = $"{context.Uri.PathAndQuery}/graph/folder1/folder 1.1";
IFolder sharedDocumentsFolder = await context.Web.GetFolderByServerRelativeUrlAsync(sharedDocumentsFolderUrl, p => p.Files);

foreach(var file in sharedDocumentsFolder.Files.AsRequested())
{
    Console.WriteLine(file.Name);
}

```

**Reply by shabirjan** [reply]
I can try that, but I don't know how many folders are there in the Dokumenter folder, that's why I am iterating all the subfolders from that  and than checking if any sub folder contains the file and than read it. Do you have any idea why above code is not working? I have followed the documentation provided by Pnp.Core SDK

**Reply by shabirjan** [reply]
@jansenbe I was able to read the files from sub folder by making small change in the code you provided. But I would really like to know why does above code didn't worked, is there anything wrong in my code that I followed from this link: https://pnp.github.io/pnpcore/using-the-sdk/files-intro.html#enumerating-files

```
using(var pnpContext = await pnpContextFactory.CreateAsync("Default")) {
                var documentFolder = (await pnpContext.Web.Lists.GetByTitleAsync("SPMTEST", p => p.RootFolder));
                var allStoreFolders = (await  documentFolder.RootFolder.GetAsync(f => f.Folders)).Folders;
            
                foreach(var folder in allStoreFolders.AsRequested()) {

                    //  Console.WriteLine(folder.Name);
                   
                        string sharedDocumentsFolderUrl = $"{pnpContext.Uri.PathAndQuery}/Delte dokumenter/{folder.Name}";
                        IFolder sharedDocumentsFolder = await pnpContext.Web.GetFolderByServerRelativeUrlAsync(sharedDocumentsFolderUrl, p => p.Files);
                        foreach (var file in sharedDocumentsFolder.Files.AsRequested())
                        {
                            Console.WriteLine(file.Name);
                        }

                        Console.WriteLine("Folder Found");
                        

                       /* await folder.LoadAsync(f => f.Files);

                        foreach (var file in folder.Files.AsRequested())
                        {
                            var extension = Path.GetExtension(file.Name);
                            if (extension == ".txt")
                            {
                                Console.WriteLine($"Filename: {file.ServerRelativeUrl}");

                            }
                        }*/
                    
                }
            }
```

**Reply by jansenbe** [reply]
@shabirjan : good that it works. You first code looks good, but don't have time to test it now. Did you consider using Search (https://pnp.github.io/pnpcore/using-the-sdk/search-intro.html) or the FindFiles (https://pnp.github.io/pnpcore/using-the-sdk/files-intro.html#finding-files) approach? The latter only is recommended for smaller libraries as it iterates across all the content of the library

**Reply by shabirjan** [reply]
@jansenbe Actually I don't want to use search or finding files approach, as i mentioned I have no idea how many folders will be there and also no idea about the name of the file in each folder, I just need to iterate and see if there are any files with .csv extension and read it. 

I would really appreciate if you test my code when you have time, so that I can implement the functionality in proper way.

**Reply by jansenbe** [reply, marked as answer]
@shabirjan : this works for me:

```csharp
var documentFolder = (await context.Web.Lists.GetByTitleAsync("graph", p => p.RootFolder));
var allStoreFolders = (await documentFolder.RootFolder.GetAsync(f => f.Folders)).Folders;

foreach (var folder in allStoreFolders.AsRequested())
{

    //  Console.WriteLine(folder.Name);
    if (folder.Name == "Folder1")
    {
        Console.WriteLine("Folder Found");

        foreach (var file in folder.Files)
        {
            var extension = Path.GetExtension(file.Name);
            if (extension == ".docx")
            {
                Console.WriteLine($"Filename: {file.ServerRelativeUrl}");
            }
        }
    }
}
```

This however is quite similar to the findfiles code: https://github.com/pnp/pnpcore/blob/dev/src/sdk/PnP.Core/Model/SharePoint/Core/Internal/Folder.cs#L446-L484

**Reply by shabirjan** [reply]
@jansenbe Ok so the code you provided, did worked for me as well, so it seem, I don't have to use

```
 await folder.LoadAsync(f => f.Files);

foreach (var file in folder.Files.AsRequested())
```

Because when I remove it, I am able to load file. Can you tell me what does this **AsRequested** to here. I also observed that my I have 339 sub folders in my Document Library, but when I was iterating it, I only got 100 folders from it, and in order to get all the folder, I have to use  `foreach (var folder in allStoreFolders.AsRequested())`
